### ISSUE TITLE
Timing issue in configuration.

### ISSUE THREAD

**Issue by nraynaud** [copterust/mpu9250#9]
When using a DMA SPI peripheral, there is no delay between the write and the reads, this leads to some configuration error. The MPU basically never goes into running mode and never sends the interrupt.

I could not pinpoint the exact problem, so I resorted to adding a delay before every write:
```rust
    fn write(&mut self, reg: Register, val: u8, spi: &mut SPI) -> Result<(), E> {
        self.ncs.set_low();
        asm::delay(72_000_000 / 100 * 1);
        spi.write(&[reg.write_address(), val])?;
        self.ncs.set_high();
        Ok(())
    }
```

This happens wether the SPI clock is a 200kHz or at 500kHz.

I am struggling to find which exact timing condition is violated in the datasheet.

**Comment by little-arhat** [comment]
We've noticed similar issue as well, but only when compiling with `opt-level=3`. All other options worked fine for us. We've "solved" that by using `"z"` level and by adding `#[inline(never)]` (https://github.com/copterust/mpu9250/commit/3c26f9b2260723e5e7bc30e6189ffc6473477e67).  Adding memory barrier after transfer also worked, but I'm not sure whether it's reordering or timing issue. 

Does compiling with `"z"` or `2` work for you, btw?

**Comment by nraynaud** [comment]
I just checked, it happens to me with this: 
```
[profile.release]
opt-level = 2
codegen-units = 1 # better optimizations
debug = true # symbols are nice and they don't increase the size on Flash
lto = true # better optimizations

```

I will test the Z and 3.

I will check the linked bug, it might see something interesting with the logic analyser.

**Comment by nraynaud** [comment]
https://github.com/nraynaud/pill-rust-test/blob/master/src/main.rs#L322
I had to resort to that to get the last version of the plugin working on my MCU.

**Comment by little-arhat** [comment]
@nraynaud Yep, I tried something along these lines as well (but only for opt-level3, as I haven't seen any issues with lower levels). I'm still not sure whether it's an issue with non volatile write somewhere, that gets reordered or timing issue (i.e. NSS pin changes state while data transfer in progress). 

Perhaps in mpu driver we should poll for transfer status and only toggle NSS pin when it's done, but that seems sub-optimal. I also think that this low level details should be handled by HAL implementation crate, as on some boards you can delegate nss control to hardware, afaik.

**Comment by nraynaud** [comment]
for me it's also a problem with O2, and I need optimization because I really don't have a lot of slack in the timing.

I don't have an oscilloscope to check the SPI protocol against a reference.

**Comment by jamwaffles** [comment]
I've been having a similar problem in `--release` mode. I was getting `panicked at 'Failed to init MPU: InvalidDevice(0)'` in the debug console which seems like a symptom of the same problem.

Adding the following to my Cargo.toml got the device working again:

```toml
[profile.release]
opt-level = 'z'
```

**Comment by romixlab** [comment]
Seems like there is similar issue in __calibrate_at_rest_ function, see the picture.
<img width="667" alt="Screen Shot 2019-04-30 at 09 48 58" src="https://user-images.githubusercontent.com/6066470/56945118-9db40880-6b2e-11e9-9083-e4c406297cc7.png">

Maybe PLL could lock that fast, or some other process inside mpu rejects transaction. With 1ms delay everything works just fine.

**Comment by little-arhat** [comment]
@romixlab what version do you use? I've added noinline attribute to write, but it seems we should just use compiler or memory barrier around ncs manipulations :-( 

Btw, I think calibrate_at_rest should be deprecated, or reimplemented.

**Comment by romixlab** [comment]
I'm using 0.11.0

**Comment by justacec** [comment]
I ran into this issue recently when trying to use SPI with an RPi Pico W board.  I was running into the issue when trying to use the `calibrate_at_rest` method.  I fixed it by inserting a `noop` loop right after the SPI CS de-assertions.  

As an aside, I do not see anything wrong with the current `calibrate_at_rest` method.  Yes, it does change a bunch of settings, but it does try to bring them back to previous settings at the end.
